src/tools/cephfs_mirror/Mirror.cc

```diff
@@ -590,6 +590,9 @@ void Mirror::schedule_mirror_update_task() {
 
   m_timer_task = new LambdaContext([this](int _) {
                                      m_timer_task = nullptr;
+                                     auto safe_timer = &(m_cct->lookup_or_create_singleton_object<SafeTimerSingleton>
```

Comment:
This is done in the constructor (Mirror::Mirrror()). How does initializing it here fixes the crash?

Reply:
It was a wrong fix. Please see the below coredump analysis.

(gdb) bt
#0  __pthread_getname_np (th=140586414077504, buf=0x7fdcf0c8b2e0 "", len=16) at pthread_getname.c:45
#1  0x00007fdcf687d465 in ceph::logging::Log::dump_recent() () from /usr/lib64/ceph/libceph-common.so.2
#2  0x000055747610b97a in handle_oneshot_fatal_signal (signum=11) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/global/signal_handler.cc:351
#3  <signal handler called>
#4  ___pthread_mutex_lock (mutex=0xb0) at pthread_mutex_lock.c:81
#5  0x00005574760d877e in __gthread_mutex_lock (__mutex=0xb0) at /usr/include/c++/11/x86_64-redhat-linux/bits/gthr-default.h:749
#6  std::mutex::lock (this=0xb0) at /usr/include/c++/11/bits/std_mutex.h:100
#7  std::scoped_lock<std::mutex>::scoped_lock (__m=..., this=<optimized out>, this=<optimized out>, __m=...) at /usr/include/c++/11/mutex:655
#8  cephfs::mirror::MirrorWatcher::is_failed (this=0x0) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/tools/cephfs_mirror/MirrorWatcher.h:56
#9  cephfs::mirror::FSMirror::is_failed (this=0x557478284340) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/tools/cephfs_mirror/FSMirror.h:52
#10 cephfs::mirror::Mirror::update_fs_mirrors (this=0x5574773c3e60) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/tools/cephfs_mirror/Mirror.cc:515
#11 0x00005574760c7f8d in Context::complete (this=0x5574774ab5c0, r=<optimized out>) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/include/Context.h:99
#12 0x00007fdcf666627e in CommonSafeTimer<std::mutex>::timer_thread() () from /usr/lib64/ceph/libceph-common.so.2
#13 0x00007fdcf666a5b1 in CommonSafeTimerThread<std::mutex>::entry() () from /usr/lib64/ceph/libceph-common.so.2
#14 0x00007fdcf5c9f802 in start_thread (arg=<optimized out>) at pthread_create.c:443
#15 0x00007fdcf5c3f450 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
=======================================
#8  cephfs::mirror::MirrorWatcher::is_failed (this=0x0) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/tools/cephfs_mirror/MirrorWatcher.h:56
56	    std::scoped_lock locker(m_lock);
(gdb) up
#9  cephfs::mirror::FSMirror::is_failed (this=0x557478284340) at /usr/src/debug/ceph-19.0.0-1578.g3d482f42.el9.x86_64/src/tools/cephfs_mirror/FSMirror.h:52
52	           m_mirror_watcher->is_failed()
(gdb) p m_mirror_watcher
$2 = (cephfs::mirror::MirrorWatcher *) 0x0

@vshankar That's^ the reason.

Reply:
EDIT : updated the text formatting.

Reply:
From the logs, I see that FSMirror is getting destroyed (destroys m_mirror_watcher too) and then init'ing again. But even when there's a valid fs_mirror in m_mirror_actions, still m_mirror_watcher = 0x0 found. See below. So it seems the bug to be fixed in `FSMirror::is_failed()` function. Any better suggestions @vshankar ?

```
(gdb) p mirror_action.fs_mirror
$1 = std::unique_ptr<cephfs::mirror::FSMirror> = {get() = 0x557478284340}
(gdb) p *mirror_action.fs_mirror
$2 = {m_cct = 0x5574773da000, m_filesystem = {fscid = 58, fs_name = "cephfs"}, m_pool_id = 116, m_service_daemon = 0x557477377380, 
  m_args = std::vector of length 4, capacity 4 = {0x7ffdad529c51 "--cluster", 0x7ffdad529c5b "ceph", 0x7ffdad529c60 "--id", 
    0x7ffdad529c65 "mirror"}, m_work_queue = 0x55747735cc00, m_lock = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = 2, 
          __count = 0, __owner = 43954, __nusers = 1, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, 
            __next = 0x0}}, __size = "\002\000\000\000\000\000\000\000\262\253\000\000\001", '\000' <repeats 26 times>, 
        __align = 2}}, <No data fields>}, m_snap_listener = {<cephfs::mirror::InstanceWatcher::Listener> = {
      _vptr.Listener = 0x557476132e88 <vtable for cephfs::mirror::FSMirror::SnapListener+16>}, fs_mirror = 0x557478284340}, 
  m_directories = std::set with 0 elements, m_all_peers = std::set with 0 elements, m_peer_replayers = std::map with 0 elements, 
  m_cluster = std::shared_ptr<librados::v14_2_0::Rados> (use count 1, weak count 0) = {get() = 0x5574773f7260}, 
  m_addrs = "172.21.15.134:0/2248701012", m_ioctx = {io_ctx_impl = 0x5574ae4d69a0}, m_instance_watcher = 0x55747977c100, 
  m_mirror_watcher = 0x0, m_retval = 0, m_stopping = false, m_init_failed = false, m_on_init_finish = 0x5574ae5d19a0, 
  m_on_shutdown_finish = 0x0, m_asok_hook = 0x5574ae556a00, m_mount = 0x557498080e60}
(gdb) p *mirror_action.fs_mirror->m_mirror_watcher
Cannot access memory at address 0x0

```

Reply:
Fixed. Please check

Reply:
> From the logs, I see that FSMirror is getting destroyed (destroys m_mirror_watcher too) and then init'ing again.

What was the case for this happening? Did initialization of FSMirror fail (m_init_failed)?

Reply:
Will check that.

Reply:
https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L584
@vshankar once the restart is completed, update_fs_mirrors() should return right? But it's scheduling again. Why?

Reply:
> https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L584 @vshankar once the restart is completed, update_fs_mirrors() should return right? But it's scheduling again. Why?

That's the timer task which periodically processes events for mirroring (enable, disable, etc..). What I'm failing to understand is the case where FSMirror:m_init_failed would be false but m_mirroring_watcher and/or m_instance_watcher will be a nullptr.

Reply:
So, it's a race b/w FSMirror reconnecting and the check for FSMirror::is_failed()? You should check if it suffices to check if a reconnect is in progress and skip the is_failed() check.

Reply:
sorry, there's a correction. I've taken another look.

```
 144291 2024-03-02T01:13:51.121+0000 7fdcf1493640 10 cephfs::mirror::Utils connect: using mon addr=172.21.15.134
 144292 2024-03-02T01:13:51.124+0000 7fdcf1493640 -1 asok(0x5574a9138000) AdminSocketConfigObs::init: failed: AdminSocket::bind_and_l\
        isten: failed to bind the UNIX domain socket to '/var/run/ceph/cephfs-mirror.asok': (17) File exists
 144293 2024-03-02T01:13:51.125+0000 7fdcf1493640 10 cephfs::mirror::Utils connect: connected to cluster=ceph using client=client.mir\
        ror
 144294 2024-03-02T01:13:51.127+0000 7fdcf1493640 -1 cephfs::mirror::FSMirror init: error accessing local pool (id=112): (2) No such \
        file or directory
 144295 2024-03-02T01:13:51.127+0000 7fdcf1493640 20 cephfs::mirror::FSMirror ~FSMirror
```
This^ is the error during connect. But the connect succeeds it seems. 

The issue is:
```
r = m_cluster->ioctx_create2(m_pool_id, m_ioctx);
```
Here^ r < 0, repeatedly fails for sometime and sets `m_init_failed` to `true` and then when it succeeds later on:
```
 144296 2024-03-02T01:13:51.127+0000 7fdcf1493640 10 cephfs::mirror::Mirror enable_mirroring: starting FSMirror: filesystem={fscid=58\
        , fs_name=cephfs}
 144297 2024-03-02T01:13:51.127+0000 7fdcf1493640 10 cephfs::mirror::ServiceDaemon: 0x557477377380 add_or_update_fs_attribute: fscid=\
        58
 144298 2024-03-02T01:13:51.127+0000 7fdcf1493640 10 cephfs::mirror::ServiceDaemon: 0x557477377380 schedule_update_status
 144299 2024-03-02T01:13:51.127+0000 7fdcf1493640 20 cephfs::mirror::FSMirror init
 144300 2024-03-02T01:13:51.127+0000 7fdcf1493640 20 cephfs::mirror::Utils connect: connecting to cluster=ceph, client=client.mirror,\
         mon_host=
 144301 2024-03-02T01:13:51.140+0000 7fdcf1493640 10 cephfs::mirror::Utils connect: using mon addr=172.21.15.134
 144302 2024-03-02T01:13:52.128+0000 7fdcecc8a640 20 cephfs::mirror::ServiceDaemon: 0x557477377380 update_status: 1 filesystem(s)
 144303 2024-03-02T01:13:53.108+0000 7fdcf0c92640 20 cephfs::mirror::Mirror update_fs_mirrors
 144304 2024-03-02T01:14:01.311+0000 7fdcf1493640 10 cephfs::mirror::Utils connect: connected to cluster=ceph using client=client.mir\
        ror
 144305 2024-03-02T01:14:01.313+0000 7fdcf1493640 20 cephfs::mirror::Utils mount: filesystem={fscid=58, fs_name=cephfs}
 144306 2024-03-02T01:14:01.320+0000 7fdcf1493640 10 cephfs::mirror::Utils mount: mounted filesystem={fscid=58, fs_name=cephfs}
 144307 2024-03-02T01:14:01.320+0000 7fdcf1493640 10 cephfs::mirror::FSMirror init: rados addrs=172.21.15.134:0/2248701012
 144308 2024-03-02T01:14:01.320+0000 7fdcf1493640 20 cephfs::mirror::FSMirror init_instance_watcher
 144309 2024-03-02T01:14:01.320+0000 7fdcf1493640 20 cephfs::mirror::InstanceWatcher init
 144310 2024-03-02T01:14:01.320+0000 7fdcf1493640 20 cephfs::mirror::InstanceWatcher create_instance
 144311 2024-03-02T01:14:01.320+0000 7fdcf1493640 20 cephfs::mirror::Mirror handle_enable_mirroring: filesystem={fscid=56, fs_name=ce\
        phfs}, peers=, r=-2
 144312 2024-03-02T01:14:01.320+0000 7fdcf3c98640 -1 asok(0x5574773da000) AdminSocket: error writing response length (32) Broken pipe
 144313 2024-03-02T01:14:01.320+0000 7fdcf0c92640 -1 *** Caught signal (Segmentation fault) **
```
`m_init_failed` is initialized to `false` by default and crashes.

Reply:
> So, it's a race b/w FSMirror reconnecting and the check for FSMirror::is_failed()? You should check if it suffices to check if a reconnect is in progress and skip the is_failed() check.

@vshankar Fixed. Please check.